In [5]:
# Analysis of Disulfide Bonds in Proteins of Known Structure
# Author: Eric G. Suchanek, PhD.
# Last revision: 7/1/23 -egs-
# Cα Cβ Sγ

%matplotlib inline

import matplotlib # type: ignore
import matplotlib.pyplot as plt # type: ignore

import plotly_express as px # type: ignore


import numpy as np # type: ignore
from proteusPy import Disulfide, DisulfideList, Load_PDB_SS
from proteusPy.ProteusGlobals import *

import pandas as pd # type: ignore
import pyvista as pv # type: ignore
from pyvista import set_plot_theme # type: ignore

# print(pv.Report())
# plt.style.use("dark_background")

pv.set_jupyter_backend("trame")

#set_plot_theme("dark")

# Analysis of Disulfide Bonds in Proteins Within the RCSB Protein Data Bank
*Eric G. Suchanek, PhD. (suchanek@mac.com)* <br> <br>

## Summary
I describe the results of a structural analysis of Disulfide bonds contained in 35,819 proteins within the RCSB Protein databank, https://www.rcsb.org. These protein structures contained 120,697 Disulfide Bonds.  The analysis utilizes Python functions from my ``ProteusPy`` package https://github.com/suchanek/proteusPy/, which is built upon the excellent ``BioPython`` library (https://www.biopython.org). 

## Background
This work represents a reprise of my original Disulfide modeling analysis conducted in 1986 ([publications](#publications) item 1) as part of my doctoral dissertation. Given the original Disulfide database contained only 2xx Disulfide Bonds I felt it would be interesting to revisit the RCSB and mine the thousands of new structures. The analysis would not have been possible withoug the creation and now ongoing development of ```proteusPy```, which represents a true, modern object-oriented rewrite of my original ```proteus``` code base. The ```C``` implemntation still exists and is available at https://github.com/suchanek/proteus/.

The initial results are shown and described in the cells below.

## Intended Audience
This notebook is intended primarily for people interested in structural biophysics with an emphasis on the analysis of protein structures. As a result, it is fairly advanced. I have made an effort to describe the analysis plainly, but I assume the reader has a basic understanding of the elements of protein structure.

## Requirements
 - My Biopython fork or my delta applied, available at: https://github.com/suchanek/biopython/
 - proteusPy: https://github.com/suchanek/proteusPy/

## Introduction
Disulfide bonds are important covalent stabilizing elements in proteins, and function as intra-molecular cross-bridges. They are formed when two Sulphur-containing Cysteine (Cys) amino acid residues are close enough and in the correct geometry to form a S-S covalent bond with their terminal sidechain $S_\gamma$ atoms. The resulting residue is known as *Cystine*, or more commonly a *Disulfide* bond. This cross-bridge stabilizes the connected protein backbone via the alpha carbon, ($C_\alpha$) backbone atoms with this strong S-S covalent bond. This bond has high energy barriers to rotation, and as such has manifests partial double-bond character. Disulfide bonds most commonly occur between alpha helices and greatly enhance a protein's stability to denaturation. 


To load the disulfide database one uses the ``Load_PDB_SS()`` function as shown below:

In [6]:
# default parameters will read from the package itself.
PDB_SS = Load_PDB_SS(verbose=True, subset=False, cutoff=-1.0)
PDB_SS.describe()

-> load_PDB_SS(): Reading /Users/egs/miniforge3/envs/proteusPy/lib/python3.11/site-packages/proteusPy/data/PDB_SS_ALL_LOADER.pkl... 
-> load_PDB_SS(): Done Reading /Users/egs/miniforge3/envs/proteusPy/lib/python3.11/site-packages/proteusPy/data/PDB_SS_ALL_LOADER.pkl... 
    =========== RCSB Disulfide Database Summary ==============
       =========== Built: 2024-11-17 15:56:50 ==============
PDB IDs present:                    36968
Disulfides loaded:                  175277
Average structure resolution:       2.19 Å
Lowest Energy Disulfide:            2q7q_75D_140D
Highest Energy Disulfide:           6vxk_801B_806B
Cα distance cutoff:                 -1.00 Å
Total RAM Used:                     44.26 MB.
    ================= proteusPy: 0.97.16 =======================


In [7]:
best = PDB_SS.get_by_name("2q7q_75D_140D")
print(best)
worst = PDB_SS.get_by_name("6vxk_801B_806B")
print(worst)
duo = DisulfideList([best, worst], "bw")
duo.display(style="sb", light=False)

<Disulfide 2q7q_75D_140D, Source: 2q7q, Resolution: 1.0 Å>
<Disulfide 6vxk_801B_806B, Source: 6vxk, Resolution: 3.0 Å>


Widget(value='<iframe src="http://localhost:49851/index.html?ui=P_0x16c1f29d0_0&reconnect=auto" class="pyvista…

## Examine the Disulfide $C_\alpha-C_\alpha$ Distances

The disulfide bond maximum $C_\alpha-C_\alpha$ distance is constrained by the bond lengths and bond angles of the cystine (Cys-Cys) sidechain atoms. If the bond angles were linear the maximum possible $C_\alpha-C_\alpha$ distance would be 8.82 $\AA$. There were a 145 physically impossible disulfide bonds in the original database who had $C_\alpha-C_\alpha$ distances > 8.0 $\AA$; these were removed. The current database therefore reflects all disulfide bonds  <= 8.0 $\AA$.

In [8]:
# retrieve the torsions dataframe
from proteusPy.Disulfide import Torsion_DF_Cols

_SSdf = PDB_SS.getTorsions()

# there are a few structures with bad SSBonds. Their
# CA distances are > 7.0. We remove them from consideration
# below

_far = _SSdf["ca_distance"] > 8.0
_near = _SSdf["ca_distance"] <= 8.0

SS_df_Far = _SSdf[_far]

# entire database with near cutoff of 8.0
SS_df = _SSdf[_near]
SS_df = SS_df[Torsion_DF_Cols].copy()

In [9]:
SS_df_Far.describe()

proximal       distal        chi1        chi2        chi3  \
count   559.000000   559.000000  559.000000  559.000000  559.000000   
mean    125.254025   201.686941  -35.786680  -33.447585  -19.541439   
std     173.670361   502.760324   87.434055  112.301395  102.168850   
min       1.000000     1.000000 -179.824618 -179.835179 -179.550141   
25%      42.000000    58.000000  -80.795742 -140.369562 -122.336600   
50%      81.000000   100.000000  -59.042560  -56.595653  -27.059034   
75%     168.000000   220.000000   44.766657   57.041441   71.188634   
max    1652.000000  8774.000000  179.681399  179.388357  179.818781   

             chi4        chi5      energy  ca_distance  cb_distance  phi_prox  \
count  559.000000  559.000000  559.000000   559.000000   559.000000       0.0   
mean   -20.142526  -33.104353    7.762586    24.395008    24.018799       NaN   
std    113.916139   88.211107    3.229279    23.720130    24.367572       NaN   
min   -179.504810 -179.807892    0.766668     8.006992     3.418274       NaN   
25%   -115.179800  -82.442525    5.352186     9.318608     8.864119       NaN   
50%    -42.876136  -52.457844    7.201096    12.464389    13.109091       NaN   
75%    100.813302    3.879540   10.438924    29.093699    28.142767       NaN   
max    179.414787  179.742050   16.024361   159.121540   158.563242       NaN   

       psi_prox  phi_dist  psi_dist  torsion_length         rho  
count       0.0       0.0       0.0      559.000000  559.000000  
mean        NaN       NaN       NaN      230.256333    1.586767  
std         NaN       NaN       NaN       51.288152   11.625101  
min         NaN       NaN       NaN       77.516159  -47.674620  
25%         NaN       NaN       NaN      197.109920   -0.868227  
50%         NaN       NaN       NaN      237.256088    0.044441  
75%         NaN       NaN       NaN      264.142437    1.703511  
max         NaN       NaN       NaN      373.836676  133.058313

This shows that there are 559 Disulfide Bonds in the overall database with Ca-Ca distances > 8.0A. These are physically impossible, and should be removed from consideration when database statistics are considered.

In [10]:
SS_df.describe()

proximal         distal           chi1           chi2  \
count  174718.000000  174718.000000  174718.000000  174718.000000   
mean      228.508671     276.237955     -45.323241      -5.455498   
std       323.997741     326.600812     104.306432     108.945474   
min       -86.000000     -62.000000    -179.999869    -179.999596   
25%        42.000000      94.000000     -92.973340     -87.194159   
50%       134.000000     188.000000     -64.328537     -54.484086   
75%       293.000000     347.000000     -38.890149      96.286026   
max      8025.000000    9004.000000     179.999973     179.998744   

                chi3           chi4           chi5         energy  \
count  174718.000000  174718.000000  174718.000000  174718.000000   
mean       -3.991778     -22.945916     -30.345911       3.722893   
std        94.090391     111.408815      99.688961       2.482426   
min      -179.981757    -179.999698    -179.999497       0.491737   
25%       -87.916102     -97.209995     -74.968578       1.963676   
50%       -64.928453     -66.263317     -59.681106       3.182716   
75%        92.888151      82.525007      56.486687       4.579841   
max       179.994824     179.999995     179.999702      19.426512   

         ca_distance    cb_distance  phi_prox  psi_prox  phi_dist  psi_dist  \
count  174718.000000  174718.000000       0.0       0.0       0.0       0.0   
mean        5.632867       3.871891       NaN       NaN       NaN       NaN   
std         0.785516       1.121285       NaN       NaN       NaN       NaN   
min         2.706248       2.233098       NaN       NaN       NaN       NaN   
25%         5.160729       3.688931       NaN       NaN       NaN       NaN   
50%         5.712830       3.845634       NaN       NaN       NaN       NaN   
75%         6.264304       4.023624       NaN       NaN       NaN       NaN   
max         7.982302     337.252323       NaN       NaN       NaN       NaN   

       torsion_length            rho  
count   174718.000000  174718.000000  
mean       233.002946      -0.052702  
std         56.642577       7.645731  
min         81.970350    -179.997832  
25%        185.263368      -1.085389  
50%        232.009393       0.017517  
75%        272.349845       1.081389  
max        383.768927     176.781353

In [11]:
# The distances are in the SS_df dataframe and can be sorted easily
SS_df.sort_values(by=["ca_distance"], ascending=False, inplace=True)

SS_df.head(10)

source           ss_id  proximal  distal        chi1        chi2  \
51314    5nx1  5nx1_191A_220A       191     220 -166.234445   62.772563   
171962   1uto  1uto_191A_220A       191     220 -164.465203  113.955483   
58235    5nx3  5nx3_191A_220A       191     220 -151.704491   49.672133   
97894    5fmf  5fmf_2712_3022       271     302 -166.231238 -158.220724   
34005    6mkn     6mkn_9D_26D         9      26 -101.609084    2.336990   
83472    1ile  1ile_181A_392A       181     392 -175.823255  151.950393   
87428    1frv   1frv_65D_530D        65     530  172.534007  157.591115   
35794    1utp  1utp_191A_220A       191     220 -157.952677  115.502969   
170766   5nfy  5nfy_452D_477D       452     477  -61.702821  176.667372   
16114    7aoe    7aoe_27L_41L        27      41   41.067554  174.972167   

              chi3        chi4        chi5     energy  ca_distance  \
51314    69.246416 -175.116839    6.343984   5.378997     7.982302   
171962  111.606573  150.357103   -1.550059   9.643133     7.980686   
58235    86.243461 -146.703389  -25.498157   6.116053     7.980171   
97894   -61.093229  -39.013852   64.857450   3.326667     7.977441   
34005   156.620921   33.845363 -100.079143  15.235831     7.976324   
83472   177.127990  151.852724   43.537089   9.536186     7.973511   
87428   133.845969  145.930141  176.808423   6.411301     7.966823   
35794    -6.059560  -91.600359    9.064000  16.118169     7.966357   
170766  111.155123  134.971049  -57.837155   3.792630     7.965060   
16114   146.810528  158.894330  172.611206   7.240146     7.960842   

        cb_distance  phi_prox  psi_prox  phi_dist  psi_dist  torsion_length  \
51314      3.446228       NaN       NaN       NaN       NaN      258.989391   
171962     6.973190       NaN       NaN       NaN       NaN      274.044418   
58235      3.762694       NaN       NaN       NaN       NaN      234.715888   
97894      5.839774       NaN       NaN       NaN       NaN      249.254035   
34005      6.466881       NaN       NaN       NaN       NaN      214.525811   
83472      5.543745       NaN       NaN       NaN       NaN      332.162297   
87428      5.725588       NaN       NaN       NaN       NaN      353.658749   
35794      7.136947       NaN       NaN       NaN       NaN      216.331884   
170766     5.417894       NaN       NaN       NaN       NaN      262.557384   
16114      5.635119       NaN       NaN       NaN       NaN      329.995713   

             rho  
51314  -0.013230  
171962  3.185933  
58235  -0.105943  
97894  -0.577896  
34005   0.294281  
83472   2.366007  
87428   3.916042  
35794   3.204235  
170766 -0.275449  
16114  -0.260208

We can extract and visualize the four longest structures and display them, as shown below. The individual windows display the disulfide bond in 'split-bond' style, where half of the bond is colored by the respective atom color. The window title indicates the approximate torsional energy (E kcal/mol), $C_\alpha-C_\alpha$ distance ($\AA$), and the torsion length (degrees). As is apparent, the $C_\beta-S_\gamma$ bond angles are almost linear. This suggests that the disulfide might not actually exist covalently, and could reflect errors in the original model.

In [ ]:
# extract and display the 4 longest SS bonds

i = 0
quattro = DisulfideList([], "L4")

for i in range(4):
    ssid = SS_df.iloc[i]["ss_id"]
    ss = PDB_SS[ssid]
    quattro.append(ss)

quattro.display(style="sb", light=False)

Plotting $C_\alpha-C_\alpha$ distances

In [ ]:
#
labels = {"value": "Cα Distance", "variable": "Count"}
cols = ["ca_distance"]
data = SS_df[cols].copy()
px.histogram(SS_df["ca_distance"], labels=labels, template="plotly_dark")
# px.histogram(SS_df["ca_distance"], labels=labels)

## Examining Disulfide Energies
The approximate bond torsional energy (kcal/mol) of a given disulfide bond 
is calculated using the following equation: \n

$ E_{kcal/mol} \approx 2.0 * cos(3.0 * \chi_{1}) + cos(3.0 * \chi_{5}) + cos(3.0 * \chi_{2}) + $
$ cos(3.0 * \chi_{4}) + 3.5 * cos(2.0 * \chi_{3}) + 0.6 * cos(3.0 * \chi_{3}) + 10.1 $

The equation embodies the typical 3-fold rotation barriers associated with single bonds,
($\chi_1, \chi_5, \chi_2, \chi_4$) and a high 2-fold barrier for $\chi_3$, resulting from the partial double bond
character of the S-S bond. This property leads to two major disulfide families, characterized
by the sign of $\chi_3$. *Left-handed* disulfides have $\chi_3$ < 0° and *right-handed* disulfides have
$\chi_3$ > 0°.


In [ ]:
labels = {"value": "Energy", "variable": "Count"}
cols = ["energy"]
px.histogram(
    SS_df["energy"], labels=labels, histnorm="probability", template="plotly_dark"
)
# px.histogram(SS_df["energy"], labels=labels, histnorm="probability")

### Find the lowest and highest energy disulfides and display them

In [ ]:
All_SS_list = PDB_SS.SSList
ssMin, ssMax = PDB_SS.SSList.minmax_energy
minmaxlist = DisulfideList([ssMin, ssMax], "mm")
minmaxlist.display(style="cpk", light=False)

In [ ]:
minmaxlist.display_overlay(light=False)
ssMin.pprint()

Very interesting result! Even though both disulfides span almost the exact same distance 5.5 $\AA$ vs 5.2 $\AA$ they do so with entirely different geometries. The minimum energy structure's torsion length of $145.62 \AA$ is significantly shorter than the maximum's $215.61 \AA$, despite the latter's *cis* $S_\gamma - S_\gamma$ bond. This *cis* conformation sits squarely on the disulfide's energy maximum, a fact reflected in the high 18.12 kcal/mol computed torsional energy.

On the other hand, *2q7q 75D-140D*, the lowest-energy disulfide is beautiful. Its conformation (-59.358°, -59.278°, -83.663°, -59.819° -59.914°) is *at* the energy minimum (0.4889 kcal/mol) calculated by the `scipy` `minimize` module.

In [ ]:
ssMin.display(style="bs", light=False)

The structure above shows that the previous and next residues have missing atoms (the long vectors drawn).

In [ ]:
# SS_df = _SSdf[_near]
# SS_df = SS_df[Torsion_DF_Cols]

SS_df = SS_df.sort_values(by=["energy"])
ssid_list = SS_df["ss_id"].values

good_SS_list = DisulfideList([], "l")
bad_SS_list = DisulfideList([], "h")

ss = Disulfide("tmp")

# first 12 are lowest energy
for i in range(12):
    ssid = ssid_list[i]
    ss = PDB_SS[ssid]
    good_SS_list.append(ss)

for i in range(13):
    if i == 0:
        continue
    ssid = ssid_list[-i]
    ss = PDB_SS[ssid]
    bad_SS_list.append(ss)

In [ ]:
bad_SS_list.display_overlay(light=True)

bad_SS_list.display(style='sb')


In [ ]:
labels = {"value": "Distance", "variable": "Distance"}
cols = ["ca_distance", "energy"]
data = SS_df[cols].copy()
data.set_index("ca_distance")

# px.histogram(SS_df, x='ca_distance', y='torsion_length', labels=labels, histfunc='avg', nbins=90, template='plotly_dark')
px.histogram(
    SS_df,
    x="ca_distance",
    y="torsion_length",
    labels=labels,
    histfunc="avg",
    nbins=90,
)

## Examine the Disulfide Torsions
The disulfide bond's overall conformation is defined by the sidechain dihedral angles $\chi_{1}$-$\chi_{5}$. Since the S-S bond has electron delocalization, it exhibits some double-bond character with strong minima at $+90°$ and $-90°$. The *Left-handed* Disulfides have $\chi_{3}$ < 0.0° and the *Right-handed* have a $\chi_{3}$ > 0.0°.

These torsion values along with the approximate torsional energy are stored in the DisulfideLoader() class and individually within each Disulfide object. We access them via the ``DisulfideList.getTorsions()`` function.

We can get a quick look at their overall statistics using the ``Pandas.describe()`` function.



In [ ]:
labels = {"value": "Angle", "variable": "Torsion"}

px.histogram(
    SS_df,
    x="chi1",
    y="chi5",
    labels=labels,
    nbins=90,
    histfunc="avg",
    # template='plotly_dark'
)

In [ ]:
labels = {"value": "Angle", "variable": "Torsion"}

px.histogram(
    SS_df,
    x="chi2",
    y="chi4",
    labels=labels,
    nbins=90,
    histfunc="avg",
    # template='plotly_dark'
)

### Examining torsions by Disulfide Handedness
We split the dataset into these two families easily with Pandas.

In [ ]:
# make two dataframes containing left handed and right handed disulfides

_left = SS_df["chi3"] <= 0.0
_right = SS_df["chi3"] > 0.0

# left handed and right handed torsion dataframes
SS_df_Left = SS_df[_left]
SS_df_Right = SS_df[_right]

# routine creates 2 lists  for left-handed and right-handed disulfides

ss_list = PDB_SS.SSList
left_handed_SS = DisulfideList([], "left_handed")
right_handed_SS = DisulfideList([], "right_handed")

for ss in ss_list:
    if ss.chi3 < 0:
        left_handed_SS.append(ss)
    else:
        right_handed_SS.append(ss)

print(f"Left Handed: {len(left_handed_SS)}, Right Handed: {len(right_handed_SS)}")

In [ ]:
SS_df_Left.describe()

In [ ]:
SS_df_Right.describe()

We can use a hexbin plot to generate the equivalent of a heatmap to explore the Correlation for the Chi1-Chi5 and Chi2-Chi4 dihedral angles. Since the disulfide bond is symmetric about Chi3 (the S-S bond), one would expect the distribution of Chi1 to be similar to Chi5, and the same with Chi2 and Chi4.

In [ ]:
# Set the figure sizes and axis limits.
DPI = 220
WIDTH = 6.0
HEIGHT = 3.0
TORMIN = -179.0
TORMAX = 180.0
GRIDSIZE = 20

In [ ]:
# !!!

x = SS_df_Left["chi1"]
y = SS_df_Left["chi5"]

x2 = SS_df_Right["chi1"]
y2 = SS_df_Right["chi5"]

plt.rcParams.update({"font.size": 8})
plt.rcParams["text.usetex"] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

ax1, ax2 = fig.subplots(1, 2, sharey=True)

fig.suptitle(r"$\chi_{1} - \chi_{5}$ Correlation", fontsize=8)

ax1.tick_params(axis="both", which="major", labelsize=6)
ax2.tick_params(axis="both", which="major", labelsize=6)

ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(TORMIN - 1, TORMAX + 1)

ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax1.set_xlabel(r"$\chi_{1 (left-handed)}$")
ax1.set_ylabel(r"$\chi_{5}$")

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap="nipy_spectral")

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(TORMIN - 1, TORMAX + 1)

ax2.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax2.set_xlabel(r"$\chi_{1 (right-handed)}$")

ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap="nipy_spectral")

plt.show()

The distributions look extremely similar for both the left-handed and right-handed populations. Let's look at Chi2 and Chi4 to see if this holds true.

In [ ]:
x = SS_df_Left["chi2"]
y = SS_df_Left["chi4"]

x2 = SS_df_Right["chi2"]
y2 = SS_df_Right["chi4"]

plt.rcParams.update({"font.size": 8})
plt.rcParams["text.usetex"] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r"$\chi_{2} - \chi_{4}$ Correlation", fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis="both", which="major", labelsize=6)
ax2.tick_params(axis="both", which="major", labelsize=6)

ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(TORMIN - 1, TORMAX + 1)
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax1.set_xlabel(r"$\chi_{2 (left-handed)}$")
ax1.set_ylabel(r"$\chi_{4}$")

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap="nipy_spectral")

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax2.set_xlabel(r"$\chi_{2 (right-handed)}$")
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap="nipy_spectral")

plt.show()

The distributions for left/right show distinct differences. Both show the predicted minimum around -75 degrees, but the right-handed group shows a population of disfuldies in the +90 degree range.

In [ ]:
x = SS_df_Left["chi1"]
y = SS_df_Left["chi3"]

x2 = SS_df_Right["chi1"]
y2 = SS_df_Right["chi3"]

plt.rcParams.update({"font.size": 8})
plt.rcParams["text.usetex"] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r"$\chi_{1} - \chi_{3}$ Correlation", fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis="both", which="major", labelsize=6)
ax2.tick_params(axis="both", which="major", labelsize=6)

ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(TORMIN - 1, TORMAX + 1)
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax1.set_xlabel(r"$\chi_{1 (left-handed)}$")
ax1.set_ylabel(r"$\chi_{3}$")

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap="nipy_spectral")

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax2.set_xlabel(r"$\chi_{1 (right-handed)}$")
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap="nipy_spectral")

plt.show()

In [ ]:
x = SS_df_Left["chi2"]
y = SS_df_Left["chi3"]

x2 = SS_df_Right["chi2"]
y2 = SS_df_Right["chi3"]

plt.rcParams.update({"font.size": 8})
plt.rcParams["text.usetex"] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r"$\chi_{2} - \chi_{3}$ Correlation", fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis="both", which="major", labelsize=6)
ax2.tick_params(axis="both", which="major", labelsize=6)

ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(TORMIN - 1, TORMAX + 1)
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax1.set_xlabel(r"$\chi_{2 (left-handed)}$")
ax1.set_ylabel(r"$\chi_{3}$")

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap="nipy_spectral")

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax2.set_xlabel(r"$\chi_{2 (right-handed)}$")
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap="nipy_spectral")

plt.show()

In [ ]:
x = SS_df_Left["chi2"]
y = SS_df_Left["chi3"]

x2 = SS_df_Right["chi2"]
y2 = SS_df_Right["chi3"]

plt.rcParams.update({"font.size": 8})
plt.rcParams["text.usetex"] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r"$\chi_{4} - \chi_{3}$ Correlation", fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis="both", which="major", labelsize=6)
ax2.tick_params(axis="both", which="major", labelsize=6)

ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(TORMIN - 1, TORMAX + 1)
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax1.set_xlabel(r"$\chi_{4 (left-handed)}$")
ax1.set_ylabel(r"$\chi_{3}$")

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap="nipy_spectral")

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax2.set_xlabel(r"$\chi_{4 (right-handed)}$")
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap="nipy_spectral")

plt.show()

In [ ]:
#

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)
ax1 = plt.subplot(111)

x = SS_df["chi3"]
y = SS_df["energy"]

ymax = 10.0
plt.hexbin(x, y, gridsize=40, cmap="nipy_spectral")
ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(y.min(), ymax)
ax1.set_xlabel(r"$\chi_{3}$")
ax1.set_ylabel(r"Energy kcal/mol")
ax1.set_title(r"Energy vs $\chi_{3}$")
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 30))

plt.show()

In [ ]:
#
TORMAX = 180  # for left handed (Chi3 < 0)
fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)
ax1 = plt.subplot(111)

x = SS_df["chi3"]
y = SS_df["ca_distance"]

ymax = 8.0
plt.hexbin(x, y, gridsize=80, cmap="nipy_spectral")
ax1.set_xlim(TORMIN - 1, TORMAX)
ax1.set_ylim(y.min(), y.max())
ax1.set_xlabel(r"$\chi_{3}$")
ax1.set_ylabel(r"C${\alpha}$ Distance ($\AA$)")
ax1.set_title(r"C${\alpha}$ Distance vs $\chi_{3}$")
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 30))

plt.show()

## Disulfide Clustering
If we could cluster the disulfides by their conformation we would be able to find conformational families; common structural motifs the recur within protein structures. The ```Scikit-Learn``` package offers a variety of clustering techniques. I explore some of them in the cells below.

### Gaussian Mixture

In [ ]:
from sklearn.mixture import GaussianMixture

n_clusters = 8

# _cols = ['chi1', 'chi2', 'chi3', 'chi4', 'chi5', 'torsion_length', 'energy', 'ca_distance']
_cols = ["chi1", "chi2", "chi3", "energy", "torsion_length"]

tor_df = SS_df[_cols]
tor_df.head(1)
gmm_model = GaussianMixture(n_components=n_clusters)
gmm_model.fit(tor_df)
cluster_labels = gmm_model.predict(tor_df)
X = pd.DataFrame(tor_df)
X["cluster"] = cluster_labels
for k in range(n_clusters):
    data = X[X["cluster"] == k]
    plt.scatter(data["torsion_length"], data["energy"], s=1)

plt.show()

In [ ]:
tor_df.head()

### Spectral Clustering

In [ ]:
# takes over an hour for full dataset on m1 pro
from sklearn.cluster import SpectralClustering
import seaborn as sns

# _cols = ['chi1', 'chi2', 'chi3', 'chi4', 'chi5', 'torsion_length', 'energy']
_cols = ["ca_distance", "chi3", "energy", "torsion_length"]

# tor_df = SS_df[['chi1', 'chi2', 'chi3', 'chi4', 'chi5']].copy()

tor_df = SS_df[_cols].copy()

X = tor_df
n_clusters = 4

scm_model = SpectralClustering(
    n_clusters=n_clusters,
    random_state=25,
    n_neighbors=16,
    affinity="nearest_neighbors",
)
# takes 51 min with full dataset
X["cluster"] = scm_model.fit_predict(X)

fig, ax = plt.subplots()
ax.set(title="Spectral Clustering")
sns.scatterplot(y="energy", x="torsion_length", data=X, hue="cluster", ax=ax, size=2)

### Affinity Propagation

In [ ]:
# takes over an hour for full dataset on M1 pro
%%timeit

from sklearn.cluster import AffinityPropagation
import seaborn as sns
n_clusters = 6
custom_palette = sns.color_palette("Dark2", n_clusters)


_cols = ['chi1', 'chi2', 'chi3', 'chi4', 'chi5', 'torsion_length']
tor_df = SS_df[_cols].copy()

X = tor_df.copy()

aff_model = AffinityPropagation(max_iter=300, random_state=25)
# takes 51 min with full dataset
X['cluster'] = aff_model.fit_predict(X)

fig, ax = plt.subplots()
ax.set(title='Affinity Propagation')
#sns.scatterplot(x='chi1', y='chi5', data=X, hue='cluster', ax=ax, size=2, palette=custom_palette)



In [ ]:
fig, ax = plt.subplots()
ax.set(title="Affinity Propagation")
sns.scatterplot(x="chi1", y="chi5", data=X, hue="cluster", ax=ax, size=2)

### Agglomerative Clustering

In [ ]:
# takes over an hour for full dataset

from sklearn.cluster import AgglomerativeClustering

n_clusters = 4
custom_palette = sns.color_palette("Dark2", n_clusters)

tor_df = SS_df[["chi1", "chi2", "chi3", "chi4", "chi5", "torsion_length"]].copy()
X = tor_df.copy()

agg_model = AgglomerativeClustering(n_clusters=n_clusters)

X["cluster"] = agg_model.fit_predict(X[["torsion_length"]])

fig, ax = plt.subplots()
ax.set(title="Agglomerative Clustering")
sns.scatterplot(
    x="chi1",
    y="chi5",
    data=X,
    hue="cluster",
    ax=ax,
    size=2,
    palette=custom_palette,
)

In [ ]:
# takes over an hour for full dataset
%%timeit

from sklearn.cluster import AgglomerativeClustering

n_clusters = 8
custom_palette = sns.color_palette("Dark2", n_clusters)


tor_df = SS_df[
    ["chi1", "chi2", "chi3", "chi4", "chi5", "energy", "ca_distance"]
].copy()
X = tor_df.copy()

agg_model = AgglomerativeClustering(n_clusters=n_clusters)

X["cluster"] = agg_model.fit_predict(X[["energy"]])

fig, ax = plt.subplots()
ax.set(title="Agglomerative Clustering")
sns.scatterplot(
    x="chi1",
    y="energy",
    data=X,
    hue="cluster",
    ax=ax,
    size=2,
    palette=custom_palette,
)

In [ ]:
# takes over an hour for full dataset
%%timeit

from sklearn.cluster import AgglomerativeClustering

n_clusters = 8
custom_palette = sns.color_palette("Dark2", n_clusters)


tor_df = SS_df[
    ["chi1", "chi2", "chi3", "chi4", "chi5", "ca_distance", "torsion_length"]
].copy()
X = tor_df.copy()

agg_model = AgglomerativeClustering(n_clusters=n_clusters)

X["cluster"] = agg_model.fit_predict(X)

fig, ax = plt.subplots()
ax.set(title="Agglomerative Clustering")
sns.scatterplot(
    x="chi1",
    y="chi2",
    data=X,
    hue="cluster",
    ax=ax,
    size=1,
    palette=custom_palette,
)

In [ ]:
# takes over an hour for full dataset
%%timeit

from sklearn.cluster import AgglomerativeClustering
import seaborn as sns

n_clusters = 4
custom_palette = sns.color_palette("Dark2", n_clusters)


tor_df = SS_df[
    ["chi1", "chi2", "chi3", "chi4", "chi5", "torsion_length"]
].copy()
X = tor_df.copy()

agg_model = AgglomerativeClustering(n_clusters=n_clusters)

X["cluster"] = agg_model.fit_predict(X)

fig, ax = plt.subplots()
ax.set(title="Agglomerative Clustering")
sns.scatterplot(
    x="chi1",
    y="chi3",
    data=X,
    hue="cluster",
    ax=ax,
    size=2,
    palette=custom_palette,
)

In [ ]:
# chatgpt output
%%timeit

from sklearn.cluster import AgglomerativeClustering


def cluster_by_family(data, n_clusters=3):
    agg_clustering = AgglomerativeClustering(n_clusters=n_clusters)
    labels = agg_clustering.fit_predict(data)
    return labels

In [ ]:
labels = {"value": "Chi3", "variable": "Variable"}

x = SS_df["chi3"]
y = SS_df["energy"]
df = SS_df[["chi3", "energy"]].copy()

energy_hist = np.histogram2d(x, y=y, bins=360)
energy_hist = energy_hist

fig = px.histogram(df, labels=labels)

fig.show()

In [ ]:
cols = ["chi1", "chi5"]
px.histogram(SS_df[cols], labels=labels)

In [ ]:
cols = ["chi2", "chi4"]
px.histogram(SS_df[cols], labels=labels)

In [ ]:
px.histogram(SS_df["chi3"], labels=labels)

## Examining Disulfide Conformational Families
Richardson et al characterized two principal conformational families for Disulfide bonds, a *Left Handed Spiral* with dihedral angles around, -60, -90, -90, -90, -60 and a *Right Handed Hook* with dihedrals, -60, 120, 90, -50, -60. My research found another class, the *Left Handed Hook* with dihedrals x, x, x, x, x. We can easily examine the representative neighbors of these families using one of the nearest neighbor functions available through the ```Disulfide``` class.

In [ ]:
# retrieve the torsions dataframe
from proteusPy.Disulfide import Torsion_DF_Cols

_SSdf = PDB_SS.getTorsions()

# there are a few structures with bad SSBonds. Their
# CA distances are > 7.0. We remove them from consideration
# below

_near = _SSdf["ca_distance"] < 9.0

# entire database
SS_df = _SSdf[_near]

SS_df = SS_df[Torsion_DF_Cols].copy()
SS_df.describe()

sslist = PDB_SS.SSList
LHS_neighbors = sslist.nearest_neighbors(-60, -90, -90, -90, -60, 12.0)
LHS_neighbors.length

In [ ]:
LHS_neighbors.display_overlay(light=True)

In [ ]:
sslist = PDB_SS.SSList
RHH_neighbors = sslist.nearest_neighbors(-60, 120, 90, -50, -60, 20)
RHH_neighbors.length

In [ ]:
RHH_neighbors.display_overlay(light=True)

## Distribution of Backbone Dihedral Angles $\phi - \psi$
A protein's backbone conformation is defined by the dihdedral angles $\phi$ and $\psi$. The ```proteusPy.Disulfide.ExtractDisulfides()``` function computes these at time of Disulfide extraction.

In [ ]:
GRIDSIZE = 40

x = SS_df["phi_prox"]
y = SS_df["psi_prox"]

x2 = SS_df["phi_dist"]
y2 = SS_df["psi_dist"]

plt.rcParams.update({"font.size": 8})
plt.rcParams["text.usetex"] = True

fig = plt.figure(figsize=(WIDTH, HEIGHT), dpi=DPI)

fig.suptitle(r"RCSB Disulfide Bond $\phi - \psi$ Dihedral Angles", fontsize=8)

ax1, ax2 = fig.subplots(1, 2, sharey=False)

ax1.tick_params(axis="both", which="major", labelsize=6)
ax2.tick_params(axis="both", which="major", labelsize=6)

ax1.set_xlim(TORMIN - 1, TORMAX + 1)
ax1.set_ylim(TORMIN - 1, TORMAX + 1)
ax1.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax1.set_xlabel(r"$\phi$ Proximal")
ax1.set_ylabel(r"$\psi$")

ax1.hexbin(x, y, gridsize=GRIDSIZE, cmap="nipy_spectral")

ax2.set_xlim(TORMIN, TORMAX)
ax2.set_ylim(-180, 180)
ax2.xaxis.set_ticks(np.arange(TORMIN - 1, TORMAX + 1, 60))

ax2.set_xlabel(r"$\phi$ Distal")
ax2.hexbin(x2, y2, gridsize=GRIDSIZE, cmap="nipy_spectral")

plt.show()

## Summary
Conformational analysis of 294,222 disulfide bonds in 36,362 proteins contained in the RCSB confirms the predominant conformational classes first described in my initial analysis:
- Left-Handed Spiral
- Right-Handed Hook
- Left-Handed Spiral
  

# Appendix - Data Pipeline

## Downloading Disulfide-containing Structures From RCSB
I started the process by creating a query on (RCSB)[https://www.rcsb.org] searching for all proteins containing one or more disulfide bonds. The resulting 38214 PDB IDs were downloaded using the ```DownloadDisulfides``` program contained in the ```programs``` folder. This took approximately 12 hours, and occupies approximately 30GB.

## Extract the Disulfides from the PDB files
The function ``Extract_Disulfides()`` processes all the .ent files in ``PDB_DIR`` and creates two .pkl files representing the Disulfide bonds contained in the scanned directory. In addition, a .csv file containing problem IDs is written if any are found. The .pkl files are consumed by the ``DisulfideLoader`` class and are considered private. You'll see numerous warnings during the scan. Files that are unparsable are removed and their IDs are logged to the problem_id.csv file. The default file locations are stored in the file globals.py and are the used by ``DisulfideExtractor()`` in the absence of arguments passed. The Disulfide parser is very stringent and will reject disulfide bonds with missing atoms or disordered atoms.


Outputs are saved in ``DATA_DIR``:
1) ``SS_PICKLE_FILE``: The ``DisulfideList`` of ``Disulfide`` objects initialized from the PDB file scan, needed by the ``DisulfideLoader()`` class.
2) ``SS_DICT_PICKLE_FILE``: the ``Dict Disulfide`` objects also needed by the ``DisulfideLoader()`` class
3) ``PROBLEM_ID_FILE``: a .csv containining the problem ids.

In general, the process only needs to be run once for a full scan. Setting the ``numb`` argument to -1 scans the entire directory. Entering a positive number allows parsing a subset of the dataset, which is useful when debugging. Setting ``verbose`` enables verbose messages. Setting ``quiet`` to ``True`` disables all warnings.

NB: A extraction of the initial disulfide bond-containing files (> 36000 files) takes about 1.25 hours on a 2020 MacbookPro with M1 Pro chip, 16GB RAM, 1TB SSD. The resulting .pkl files consume approximately 1GB of disk space, and equivalent RAM used when loaded.

## Data Cleaning
* Parsing PDB files
* Extracting disulfides
* Removing redundant disulfides

## Load the Disulfide Data
Now that the Disulfides have been extracted and the Disulfide .pkl files have been created we can load them into memory using the DisulfideLoader() class. This class stores the Disulfides internally as a DisulfideList and a dict. Array indexing operations including slicing have been overloaded, enabling straightforward access to the Disulfide bonds, both in aggregate and by residue. After loading the .pkl files the Class creates a Pandas ``DataFrame`` object consisting of the Disulfide ID, all sidechain dihedral angles, the local coordinates for the Disulfide and the computed Disulfide bond torsional energy.

NB: Loading the data takes 3.5 minutes on my MacbookPro. Be patient if it seems to take a long time to load.

The ```Disulfide``` and ```DisulfideList``` classes include rendering capabiities using the excellent PyVista interface to the VTK package. (http://pyvista.org). The following cell displays the first Disulfide bond in the database in ball-and stick style. Atoms are colored by atom type:
- Grey = Carbon
- Blue = Nitrogen
- Red = Oxygen
- Yellow = Sulfur
- White = Previous residue carbonyl carbon and next residue amino Nitrogen. (more on this below).

We can load the database and display the first disulfide as shown below. The molecular display is *interactive*; select-drag to rotate, mousewheel to zoom. The X-Y-Z widget in the window upper right allows orientation against the X, Y and Z axes. The window title provides information about the disulfide rendered:
* Disulfide source ID
* Disulfide name, which embodies the source id, the proximal residue number and the distal residue number.
* Disulfide approximate torsional energy (kcal/mol)
* Disulfide $C_\alpha-C_\alpha$ distance $\AA$
* Disulfide torsion length (5-dimensional vector length) (degrees)



# Publications
* https://doi.org/10.1021/bi00368a023
* https://doi.org/10.1021/bi00368a024
* https://doi.org/10.1016/0092-8674(92)90140-8
* http://dx.doi.org/10.2174/092986708783330566

In [ ]:
from proteusPy import *
from proteusPy.DisulfideLoader import Load_PDB_SS
from proteusPy.DisulfideList import DisulfideList
from proteusPy.Disulfide import Disulfide

PDB_SS = Load_PDB_SS(verbose=False, subset=True)

In [ ]:
ss_list = DisulfideList([], "tmp")
sslist = PDB_SS.SSList
ssmin_enrg, ssmax_enrg = PDB_SS.SSList.minmax_energy()
low_energy_neighbors = DisulfideList([], "Neighbors")
low_energy_neighbors = ssmin_enrg.Torsion_neighbors(sslist, 10)

In [ ]:
pdb = Load_PDB_SS()

In [ ]:
pdb.describe()